libflux/src/semantic/nodes.rs

```diff
@@ -710,8 +786,54 @@ pub struct CallExpr {
 }
 
 impl CallExpr {
-    fn infer(&self, env: Environment, f: &mut Fresher) -> Result {
-        unimplemented!();
+    fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
+        // First, recursively infer every type of the children of this call expression,
+        // update the environment and the constraints, and use the inferred types to
+        // build the fields of the type for this call expression.
+        let (mut env, mut cons) = self.callee.infer(env, f)?;
+        let mut req = HashMap::new();
+        let mut pipe = None;
+        for Property {
+            key: ref mut id,
+            value: ref mut expr,
+            ..
+        } in &mut self.arguments
```

Comment:
Whoa I've never seen this kind of destructuring before in a for loop. But I think I love it.

Reply:
yeah! `..` is like `and all the rest I don't care about`.